fortitude/src/rules/correctness/error_handling.rs

```diff
@@ -297,3 +305,152 @@ fn new_branch(node: &Node) -> bool {
+/// When allocating or deallocating multiple arrays at once, the use of a `stat`
+/// parameter will permit the program to continue running even if one of the
+/// allocations or deallocations fails. However, it may not be clear which
+/// allocation or deallocation caused the error.
+///
+/// To avoid confusion, it is recommended to use separate allocate or deallocate
+/// statements for each array and check the `stat` parameters individually.
+#[derive(ViolationMetadata)]
+pub(crate) struct MultipleAllocationsWithStat {
+    alloc_type: AllocationType,
+}
+
+impl Violation for MultipleAllocationsWithStat {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let allocations = match self.alloc_type {
+            AllocationType::Allocate => "allocations",
+            AllocationType::Deallocate => "deallocations",
+        };
+        format!("'stat' parameter used with multiple {allocations}.")
+    }
+}
+
+impl AstRule for MultipleAllocationsWithStat {
+    fn check(_settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
+        let src = source.source_text();
+
+        // Check this has a stat parameter
+        let stat_node = node.kwarg("stat", src)?;
+
+        // Count allocations
+        let count = match node.kind() {
+            "allocate_statement" => count_allocations(node),
+            "deallocate_statement" => count_deallocations(node),
+            _ => return None,
+        };
+        if count <= 1 {
+            return None;
+        }
+
+        let alloc_type = AllocationType::from_node(node).ok()?;
+
+        some_vec!(Diagnostic::from_node(Self { alloc_type }, &stat_node))
+    }
+
+    fn entrypoints() -> Vec<&'static str> {
+        vec!["allocate_statement", "deallocate_statement"]
+    }
+}
+
+fn count_allocations(node: &Node) -> usize {
+    node.children_by_field_name("allocation", &mut node.walk())
+        .count()
+}
+
+fn count_deallocations(node: &Node) -> usize {
+    node.named_children(&mut node.walk())
+        .filter(|c| c.kind() == "identifier")
+        .count()
+}
+
+/// ## What does it do?
+/// This rule detects whether `stat` is used without also setting `errmsg` when
+/// allocating or deallocating. Similarly checks for the use of `iostat` without
+/// `iomsg` with IO routines, and `cmdstat` without `cmdmsg` when using
+/// `execute_command_line`.
+///
+/// ## Why is this bad?
+/// The error codes returned when using `stat`, `iostat`, or `cmdstat` are not
+/// very informative on their own, and are not portable across compilers. It is
+/// recommended to always capture the associated error message alongside the
+/// error code:
+///
+/// ```f90
+/// real, allocatable :: x(:)
+/// integer :: status
+/// character(256) :: message
+///
+/// allocate (x(100), stat=status, errmsg=message)
+/// open (unit=10, file="data.txt", iostat=status, iomsg=message)
+/// call execute_command_line("ls", cmdstat=status, cmdmsg=message)
+/// ```
```

Comment:
Might be worth mentioning that in F2023, these functions can automatically allocate `errmsg` to be the correct size (although I don't think anybody actually implements this yet!)